### ISSUE TITLE
The magic of DDPWrapper

### ISSUE THREAD

**Issue by twmht** [open-mmlab/mmrazor#99]
Hi,

I found out when doing accumulating gradient (https://github.com/open-mmlab/mmrazor/blob/master/mmrazor/models/algorithms/autoslim.py#L177), it must use DDPWrapper (https://github.com/open-mmlab/mmrazor/blob/master/mmrazor/core/distributed_wrapper.py). 

Otherwise it would throw

https://github.com/pytorch/pytorch/blob/ccab142197d4647e397fc492ba4e7d857fa07a6d/torch/csrc/distributed/c10d/reducer.cpp#L421

```
 "initialize_buckets must NOT be called during autograd execution."
```

DDPWrapper wraps the top module as DDP, What is the magic behind that? Why it does not throw the exception?

thank you.

**Comment by pppppM** [comment]
This problem is very complicated. 
If you want to understand the reason, maybe you need to understand the [`DistributedDataParallel`](https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel) and [`MMDistributedDataParallel`](https://github.com/open-mmlab/mmcv/blob/master/mmcv/parallel/distributed.py#L11) first.
In MMRazor，the reason is that AutoSlim's `optimizer.step()` is executed in `train_step` , not in `OptimizerHook` like other repo.

**Comment by twmht** [comment]
thank you.
